### PR TITLE
BUG: Keep float dtype in merge on int and float column

### LINKED ISSUE #16572: merge of int and float column results in column of dtype object
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

A = pd.DataFrame({'X': [1, 2, 3]})
B = pd.DataFrame({'Y': [1.0, 3.0]})

B = B.merge(A, left_on='Y', right_on='X')

print(B.dtypes)
```
#### Problem description
The output is
```
Y    object
X     int64
dtype: object
```
meaning that Y was incorrectly upcast to object.

In pandas 0.19, the dtype of B.Y after merging is float64.
In pandas 0.20, the dtype of B.Y after merging is object.

The problem only occurs if some keys in B.Y are missing (try adding 2.0)

#### Expected Output
```
Y    float64
X      int64
dtype: object
```

#### Output of ``pd.show_versions()``
------------------

<details>
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-2-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.1
pytest: None
pip: 9.0.1
setuptools: 35.0.2
Cython: None
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 6.0.0
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by reidy-p**
- [ ] closes #16572
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/reshape/test_merge.py [hunk 1]
```diff
@@ -1408,6 +1412,25 @@ def test_join_multi_dtypes(self, d1, d2):
...
+    ])
+    def test_merge_on_ints_floats(self, int_vals, float_vals):
+        # GH 16572
+        # Check that float column is not cast to object if
+        # merging on float and int
+        A = DataFrame({'X': int_vals})
+        B = DataFrame({'Y': float_vals})
+
+        res = A.merge(B, left_on='X', right_on='Y')
+        assert is_float_dtype(res['Y'].dtype)
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: you need to check the results
the last merge is not what you would expect

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/merge.py [hunk 2]
```diff
@@ -906,13 +906,20 @@ def _maybe_coerce_merge_keys(self):
...
-            # kinds to proceed, eg. int64 and int8
+            # kinds to proceed, eg. int64 and int8, int and float
             # further if we are object, but we infer to
             # the same, then proceed
             if is_numeric_dtype(lk) and is_numeric_dtype(rk):
                 if lk.dtype.kind == rk.dtype.kind:
                     continue
 
+                # check whether ints and floats
+                if is_integer_dtype(rk) and is_float_dtype(lk):
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: I am a little leary of merging integer & float columns when the floats are not actually == ints. A warning would be good here if 

``not (lk == lk.astype(rk)).all()`` is True (and the reverse for 2nd case)

File: pandas/core/reshape/merge.py [hunk 3]
```diff
@@ -906,13 +906,20 @@ def _maybe_coerce_merge_keys(self):
                 continue
 
             # if we are numeric, then allow differing
-            # kinds to proceed, eg. int64 and int8
+            # kinds to proceed, eg. int64 and int8, int and float
             # further if we are object, but we infer to
             # the same, then proceed
             if is_numeric_dtype(lk) and is_numeric_dtype(rk):
                 if lk.dtype.kind == rk.dtype.kind:
                     continue
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: also, let's change this to an ``if/elsif/else`` clause (and use pass rather than continue to delineate the cases)

**Review by jreback — CHANGES_REQUESTED** [review]
this needs a sub-section in other enhancements

File: pandas/tests/reshape/test_merge.py (hunk 1 above)
Comment (reidy-p) [reply pandas/tests/reshape/test_merge.py]: On my branch:
```
In [1]: A = pd.DataFrame({'X': [1, 2, 3]})
In [2]: B = pd.DataFrame({'Y': [1.1, 2.5, 3.0]})

In [3]: A.merge(B, left_on='X', right_on='Y'))
Out[3]: 
   X    Y
0  3  3.0

In [4]: A.merge(B, left_on='X', right_on='Y').dtypes
Out[4]:
X      int64
Y    float64
dtype: object
```
Is this not what we would expect? 

And could you give me an example of the problems caused when merging integer and float columns when the float values are not equal to the int representation of those same values? Thanks.

File: pandas/core/reshape/merge.py [hunk 4]
```diff
@@ -906,17 +906,37 @@ def _maybe_coerce_merge_keys(self):
...
+                                      'are not equal to their int '
+                                      'representation', UserWarning)
+                    continue
+
                 # let's infer and see if we are ok
-                if lib.infer_dtype(lk) == lib.infer_dtype(rk):
+                elif lib.infer_dtype(lk) == lib.infer_dtype(rk):
                     continue
 
+                else:
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: oh I meant put the 'houston we have a problem' in the else. then everything else is just a pass.

File: pandas/tests/reshape/test_merge.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/reshape/test_merge.py]: hmm that looks ok, in any event compare the expected results (not the dtypes)

Comment (jreback) [reply pandas/tests/reshape/test_merge.py]: the issue with comparing floats and integers is almost always wrong, IOW the user needs to be aware of this, as its almost always an error (so warning is ok here).

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/merge.py [hunk 5]
```diff
@@ -906,16 +906,33 @@ def _maybe_coerce_merge_keys(self):
...
+                    pass
+
+                # check whether ints and floats
+                elif is_integer_dtype(rk) and is_float_dtype(lk):
+                    if not (lk == lk.astype(rk.dtype)).all():
+                        warnings.warn('You are merging on int and float '
+                                      'columns where the float values '
+                                      'are not equal to their int '
+                                      'representation', UserWarning)
+                    pass
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: you don't actually need the pass here (as you have the if statement)

File: pandas/core/reshape/merge.py [hunk 6]
```diff
@@ -906,16 +906,33 @@ def _maybe_coerce_merge_keys(self):
...
+
+                elif is_float_dtype(rk) and is_integer_dtype(lk):
+                    if not (rk == rk.astype(lk.dtype)).all():
+                        warnings.warn('You are merging on int and float '
+                                      'columns where the float values '
+                                      'are not equal to their int '
+                                      'representation', UserWarning)
+                    pass
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: same

File: pandas/core/reshape/merge.py [hunk 7]
```diff
@@ -906,16 +906,33 @@ def _maybe_coerce_merge_keys(self):
...
 
                 # let's infer and see if we are ok
-                if lib.infer_dtype(lk) == lib.infer_dtype(rk):
-                    continue
+                elif lib.infer_dtype(lk) == lib.infer_dtype(rk):
+                    pass
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: this ones good!

File: doc/source/whatsnew/v0.22.0.txt [hunk 8]
```diff
@@ -29,6 +29,7 @@ Other Enhancements
 - :class:`pandas.io.formats.style.Styler` now has method ``hide_columns()`` to determine whether columns will be hidden in output (:issue:`14194`)
 - Improved wording of ``ValueError`` raised in :func:`to_datetime` when ``unit=`` is passed with a non-convertible value (:issue:`14350`)
 - :func:`Series.fillna` now accepts a Series or a dict as a ``value`` for a categorical dtype (:issue:`17033`)
+- :func:`pandas.DataFrame.merge` no longer casts a ``float`` column to ``object`` when merging on ``int`` and ``float`` columns (:issue:`16572`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: move to api_breaking

File: pandas/tests/reshape/test_merge.py [hunk 9]
```diff
@@ -1408,6 +1411,42 @@ def test_join_multi_dtypes(self, d1, d2):
...
+        ([1, 2, 3], [1.0, 3.0], {'X': [1, 3], 'Y': [1.0, 3.0]}),
+        ([1, 2], [1.0, 2.0, 3.0], {'X': [1, 2], 'Y': [1.0, 2.0]}),
+    ])
+    def test_merge_on_ints_floats(self, int_vals, float_vals, exp_vals):
+        # GH 16572
+        # Check that float column is not cast to object if
+        # merging on float and int columns
+        A = DataFrame({'X': int_vals})
+        B = DataFrame({'Y': float_vals})
+        exp_res = DataFrame(exp_vals)
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: call this expected

File: pandas/tests/reshape/test_merge.py [hunk 10]
```diff
@@ -1408,6 +1411,42 @@ def test_join_multi_dtypes(self, d1, d2):
...
+
+        res = A.merge(B, left_on='X', right_on='Y')
+        assert_frame_equal(res, exp_res)
+
+        res = B.merge(A, left_on='Y', right_on='X')
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: call these result

File: pandas/tests/reshape/test_merge.py [hunk 11]
```diff
@@ -1408,6 +1411,42 @@ def test_join_multi_dtypes(self, d1, d2):
...
+        assert_frame_equal(res, exp_res[['Y', 'X']])
+
+    def test_merge_on_ints_floats_warning(self):
+        # GH 16572
+        # merge will produce a warning when merging on int and
+        # float columns where the float values are not exactly
+        # equal to their int representation
+        A = DataFrame({'X': [1, 2, 3]})
+        B = DataFrame({'Y': [1.1, 2.5, 3.0]})
+        exp_res = DataFrame({'X': [3], 'Y': [3.0]})
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: same as above

**Review by jreback — CHANGES_REQUESTED** [review]
minor comments. ping on green.

**Comment by reidy-p** [comment]
@jreback thanks for the comments. Green now.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @reidy-p nice patch and great responsiveness!

**Pull request merged**